working/macros/example/data_class.dart

```diff
@@ -0,0 +1,241 @@
+          var defaultValue = param.defaultValue == null
+              ? ''
+              : Code.fromParts([' = ', param.defaultValue!]);
+          parts.add(Code.fromParts([
+            '\n$requiredKeyword${param.type.toCode()} ${param.name}',
+            defaultValue,
+            ',',
+          ]));
+        }
+      }
+      parts.add(Code.fromString('\n})'));
+      if (superconstructor != null) {
+        parts.add(Code.fromString(' : super('));
+        for (var param in superconstructor.positionalParameters) {
+          parts.add(Code.fromString('\n${param.name},'));
+        }
+        if (superconstructor.namedParameters.isNotEmpty) {
+          parts.add(Code.fromString('{'));
+          for (var param in superconstructor.namedParameters) {
+            parts.add(Code.fromString('\n${param.name}: ${param.name},'));
+          }
+          parts.add(Code.fromString('\n}'));
+        }
+        parts.add(Code.fromString(')'));
+      }
+      parts.add(Code.fromString(';'));
+      builder.addToClass(DeclarationCode.fromParts(parts));
+    });
+  }
+}
+
+const copyWith = _CopyWith();
+
+// TODO: How to deal with overriding nullable fields to `null`?
+class _CopyWith extends ClassMacro {
+  const _CopyWith();
+
+  @override
+  void visitClass(ClassDeclaration clazz) {
+    buildDeclarations((builder) async {
+      var methods = await builder.methodsOf(clazz);
+      if (methods.any((c) => c.name == 'copyWith')) {
+        throw ArgumentError(
+            'Cannot generate a copyWith method because one already exists');
+      }
+      var allFields = await clazz.allFields(builder).toList();
+      var namedParams = [
+        for (var field in allFields)
+          ParameterCode.fromString(
+              '${field.type.toCode()}${field.type.isNullable ? '' : '?'} '
+              '${field.name}'),
+      ];
+      var args = [
+        for (var field in allFields)
+          NamedArgumentCode.fromString(
+              '${field.name}: ${field.name} ?? this.${field.name}'),
+      ];
+      builder.addToClass(DeclarationCode.fromParts([
+        clazz.instantiate().toCode(),
+        ' copyWith({',
+        ...namedParams.joinAsCode(', '),
+        ',})',
+        // TODO: We assume this constructor exists, but should check
+        '=> ', clazz.instantiate().toCode(), '(',
+        ...args.joinAsCode(', '),
+        ', );',
+      ]));
+    });
+  }
+}
+
+const hashCode = _HashCode();
+
+class _HashCode extends ClassMacro {
+  const _HashCode();
+
+  @override
+  void visitClass(ClassDeclaration clazz) {
+    buildDeclarations((builder) async {
+      var hashCodeExprs = [
+        await for (var field in clazz.allFields(builder))
+          ExpressionCode.fromString('${field.name}.hashCode')
+      ].joinAsCode(' ^ ');
+
+      builder.addToClass(DeclarationCode.fromParts([
+        '''
+@override
+int get hashCode =>''',
+        ...hashCodeExprs,
+        ';',
+      ]));
+    });
```

Comment:
These look like they are both introspecting on members and declaring members in the same phase. Is that a problem? What if they need to see the declarations of fields produced by other macros? Would something like this work?

```dart
class _HashCode extends ClassMacro {
  const _HashCode();

  @override
  void visitClass(ClassDeclaration clazz) {
    // The visitClass() method body runs in phase 2...
    clazz.addMethod('@override int get hashCode', (builder)  async {
      // ...and then the macro engine defers running the callback (which has access to the builder and
      // instrospection) until phase 3:
      var hashCodeExprs = [
        await for (var field in builder.allFields(builder))
          ExpressionCode.fromString('${field.name}.hashCode')
      ].joinAsCode(' ^ ');

      return DeclarationCode.fromParts([...hashCodeExprs, ';']);
    });
  }
}
```

Reply:
> These look like they are both introspecting on members and declaring members in the same phase. Is that a problem?

That is a good point, for the ones where the signature doesn't also depend on the fields, I should split this up.

For things like `CopyWith` and `AutoConstructor` they have to use the fields to declare their signatures though. 

> ```dart
> clazz.addMethod('@override int get hashCode', (builder)  async {
> ```

I don't want to hang things off of the `ClassDeclaration` - you can get one of those for anything in the program.

I did consider an api like this, but you may also want to provide a body for something you didn't define, (or more commonly wrap something you didn't define). So we would have to provide two different apis for doing that. So that is why I went with the separate api here.